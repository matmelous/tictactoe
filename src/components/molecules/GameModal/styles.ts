export const modalBoxStyle = {
  position: "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  minWidth: 300,
  maxWidth: 550,
  width: "60%",
  bgcolor: "background.paper",
  boxShadow: 24,
  p: 4,
};
