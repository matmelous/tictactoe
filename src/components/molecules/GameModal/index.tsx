import React from "react";

import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import Modal from "@mui/material/Modal";
import { modalBoxStyle } from "./styles";
import * as language from "../../../localization/languageKeys";

type PropsType = {
  open: boolean;
  title: string;
  description: string | React.ReactNode;
  body?: React.ReactNode;
  buttonAction: () => void;
  handleClose: () => void;
  buttonDisabled?: boolean;
  buttonText?: string;
};

const GameModal = ({
  open,
  title,
  description,
  buttonAction,
  body,
  handleClose,
  buttonDisabled,
  buttonText,
}: PropsType) => {
  return (
    <Modal
      open={open}
      onClose={handleClose}
      aria-labelledby="modal-modal-title"
      aria-describedby="modal-modal-description"
      data-testid="game-modal"
    >
      <Box sx={modalBoxStyle}>
        <Typography id="modal-modal-title" variant="h6" component="h2">
          {title}
        </Typography>
        <Typography id="modal-modal-description" sx={{ mt: 2 }} variant="h4">
          {description}
        </Typography>
        {body}
        <Button
          style={{ marginTop: 35 }}
          variant="contained"
          color="primary"
          onClick={buttonAction}
          disabled={buttonDisabled}
        >
          {buttonText}
        </Button>
      </Box>
    </Modal>
  );
};

GameModal.defaultProps = {
  body: null,
  buttonDisabled: false,
  buttonText: language.PLAY_AGAIN,
};

export default GameModal;
