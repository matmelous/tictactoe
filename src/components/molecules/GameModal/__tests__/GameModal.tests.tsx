import React from "react";
import { render, cleanup, screen } from "@testing-library/react";
import GameModal from "../index";

describe("GameModal", () => {
  afterEach(cleanup);

  it("renders open modal", () => {
    const { asFragment } = render(
      <GameModal
        open={true}
        title="test title"
        description="test description"
        buttonAction={() => null}
        handleClose={() => null}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.queryByTestId("game-modal")).toBeInTheDocument();
    expect(screen.queryByText("test title")).toBeInTheDocument();
    expect(screen.queryByText("test description")).toBeInTheDocument();
  });

  it("renders closed modal", () => {
    render(
      <GameModal
        open={false}
        title="test title"
        description="test description"
        buttonAction={() => null}
        handleClose={() => null}
      />
    );
    expect(screen.queryByTestId("game-modal")).not.toBeTruthy();
    expect(screen.queryByText("test title")).not.toBeInTheDocument();
    expect(screen.queryByText("test description")).not.toBeInTheDocument();
  });

  it("renders disabled button", () => {
    render(
      <GameModal
        open={true}
        title="test title"
        description="test description"
        buttonAction={() => null}
        handleClose={() => null}
        buttonDisabled={true}
        buttonText="start"
      />
    );
    expect(screen.queryByRole("button")).toBeDisabled();
    expect(screen.queryByText("start")).toBeInTheDocument();
    expect(screen.queryByText("test title")).toBeInTheDocument();
    expect(screen.queryByText("test description")).toBeInTheDocument();
  });

  it("renders enabled button and clicks", () => {
    const buttonAction = jest.fn();

    render(
      <GameModal
        open={true}
        title="test title"
        description="test description"
        buttonAction={buttonAction}
        handleClose={() => null}
        buttonDisabled={false}
        buttonText="start"
      />
    );
    expect(screen.queryByRole("button")).not.toBeDisabled();
    expect(screen.queryByText("start")).toBeInTheDocument();
    expect(screen.queryByText("test title")).toBeInTheDocument();
    expect(screen.queryByText("test description")).toBeInTheDocument();
    screen.getByRole("button").click();
    expect(buttonAction).toHaveBeenCalledTimes(1);
  });
});
