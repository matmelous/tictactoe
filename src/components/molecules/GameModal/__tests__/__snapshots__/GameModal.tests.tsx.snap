// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders open modal 1`] = `<DocumentFragment />`;
